From 78c930e758082ed351a5e9b1d50a2515d642cb08 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Jakub=20Jank=C5=AF?= <janku.jakub.jj@gmail.com>
Date: Tue, 6 Jun 2017 23:22:37 +0200
Subject: [PATCH 36/55] file-xfer: Send more detailed error messages

Send VD_AGENT_FILE_XFER_STATUS_VDAGENT_NOT_CONNECTED,
VD_AGENT_FILE_XFER_STATUS_SESSION_LOCKED or
VD_AGENT_FILE_XFER_STATUS_DISABLED instead of general error, when the
given error occurs.

send_file_xfer_status() ensures, these errors are sent only to clients with
VD_AGENT_CAP_FILE_XFER_DETAILED_ERRORS capability.

Acked-by: Victor Toso <victortoso@redhat.com>
---
 src/vdagent/file-xfers.c |  4 ++--
 src/vdagent/file-xfers.h |  2 +-
 src/vdagent/vdagent.c    | 12 ++++++------
 src/vdagentd/vdagentd.c  |  8 ++++++--
 4 files changed, 15 insertions(+), 11 deletions(-)

diff --git a/src/vdagent/file-xfers.c b/src/vdagent/file-xfers.c
index af73d87..4d76003 100644
--- a/src/vdagent/file-xfers.c
+++ b/src/vdagent/file-xfers.c
@@ -359,10 +359,10 @@ void vdagent_file_xfers_data(struct vdagent_file_xfers *xfers,
     }
 }
 
-void vdagent_file_xfers_error(struct udscs_connection *vdagentd, uint32_t msg_id)
+void vdagent_file_xfers_error_disabled(struct udscs_connection *vdagentd, uint32_t msg_id)
 {
     g_return_if_fail(vdagentd != NULL);
 
     udscs_write(vdagentd, VDAGENTD_FILE_XFER_STATUS,
-                msg_id, VD_AGENT_FILE_XFER_STATUS_ERROR, NULL, 0);
+                msg_id, VD_AGENT_FILE_XFER_STATUS_DISABLED, NULL, 0);
 }
diff --git a/src/vdagent/file-xfers.h b/src/vdagent/file-xfers.h
index 28a71fd..3e2ed1d 100644
--- a/src/vdagent/file-xfers.h
+++ b/src/vdagent/file-xfers.h
@@ -37,7 +37,7 @@ void vdagent_file_xfers_status(struct vdagent_file_xfers *xfers,
     VDAgentFileXferStatusMessage *msg);
 void vdagent_file_xfers_data(struct vdagent_file_xfers *xfers,
     VDAgentFileXferDataMessage *msg);
-void vdagent_file_xfers_error(struct udscs_connection *vdagentd,
+void vdagent_file_xfers_error_disabled(struct udscs_connection *vdagentd,
     uint32_t msg_id);
 
 #endif
diff --git a/src/vdagent/vdagent.c b/src/vdagent/vdagent.c
index 9900303..dfb9284 100644
--- a/src/vdagent/vdagent.c
+++ b/src/vdagent/vdagent.c
@@ -89,8 +89,8 @@ static void daemon_read_complete(struct udscs_connection **connp,
             vdagent_file_xfers_start(vdagent_file_xfers,
                                      (VDAgentFileXferStartMessage *)data);
         } else {
-            vdagent_file_xfers_error(*connp,
-                                     ((VDAgentFileXferStartMessage *)data)->id);
+            vdagent_file_xfers_error_disabled(*connp,
+                                              ((VDAgentFileXferStartMessage *)data)->id);
         }
         break;
     case VDAGENTD_FILE_XFER_STATUS:
@@ -98,8 +98,8 @@ static void daemon_read_complete(struct udscs_connection **connp,
             vdagent_file_xfers_status(vdagent_file_xfers,
                                       (VDAgentFileXferStatusMessage *)data);
         } else {
-            vdagent_file_xfers_error(*connp,
-                                     ((VDAgentFileXferStatusMessage *)data)->id);
+            vdagent_file_xfers_error_disabled(*connp,
+                                              ((VDAgentFileXferStatusMessage *)data)->id);
         }
         break;
     case VDAGENTD_FILE_XFER_DISABLE:
@@ -125,8 +125,8 @@ static void daemon_read_complete(struct udscs_connection **connp,
             vdagent_file_xfers_data(vdagent_file_xfers,
                                     (VDAgentFileXferDataMessage *)data);
         } else {
-            vdagent_file_xfers_error(*connp,
-                                     ((VDAgentFileXferDataMessage *)data)->id);
+            vdagent_file_xfers_error_disabled(*connp,
+                                              ((VDAgentFileXferDataMessage *)data)->id);
         }
         break;
     case VDAGENTD_CLIENT_DISCONNECTED:
diff --git a/src/vdagentd/vdagentd.c b/src/vdagentd/vdagentd.c
index 5373909..1f10f1c 100644
--- a/src/vdagentd/vdagentd.c
+++ b/src/vdagentd/vdagentd.c
@@ -349,14 +349,14 @@ static void do_client_file_xfer(struct vdagent_virtio_port *vport,
             send_file_xfer_status(vport,
                "Could not find an agent connection belonging to the "
                "active session, cancelling client file-xfer request %u",
-               s->id, VD_AGENT_FILE_XFER_STATUS_CANCELLED, NULL, 0);
+               s->id, VD_AGENT_FILE_XFER_STATUS_VDAGENT_NOT_CONNECTED, NULL, 0);
             return;
         } else if (session_info_session_is_locked(session_info)) {
             syslog(LOG_DEBUG, "Session is locked, skipping file-xfer-start");
             send_file_xfer_status(vport,
                "User's session is locked and cannot start file transfer. "
                "Cancelling client file-xfer request %u",
-               s->id, VD_AGENT_FILE_XFER_STATUS_ERROR, NULL, 0);
+               s->id, VD_AGENT_FILE_XFER_STATUS_SESSION_LOCKED, NULL, 0);
             return;
         }
         udscs_write(active_session_conn, VDAGENTD_FILE_XFER_START, 0, 0,
@@ -932,6 +932,10 @@ static void agent_read_complete(struct udscs_connection **connp,
                                       (uint8_t*)&free_space, sizeof(uint64_t));
                 break;
             }
+            case VD_AGENT_FILE_XFER_STATUS_DISABLED:
+                send_file_xfer_status(virtio_port, "File-xfer is disabled, cancelling",
+                                      header->arg1, header->arg2, NULL, 0);
+                break;
             default:
                 send_file_xfer_status(virtio_port, NULL, header->arg1, header->arg2, NULL, 0);
         }
-- 
2.16.1

